### PR TITLE
fix(youtube/general-ads): restore swipe back to exit gesture

### PR DESCRIPTION
**Pull request by 0xrxL**

### CODE REVIEW

File: app/src/main/java/com/google/android/apps/youtube/app/watchwhile/WatchWhileActivity.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.google.android.apps.youtube.app.watchwhile;
```
Comment (oSumAtrIX) [inline app/src/main/java/com/google/android/apps/youtube/app/watchwhile/WatchWhileActivity.java:1]: This is the wrong package for dummy classes. Move this class to `https://github.com/revanced/revanced-integrations/tree/main/dummy/src/main/java/com/google/android/apps/youtube/app`.

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package app.revanced.integrations.patches;
+
+import com.google.android.apps.youtube.app.watchwhile.WatchWhileActivity;
+import app.revanced.integrations.utils.LogHelper;
+
+public class BackToExitPatch {
+    private static long backPressedCount = 0;
+
+    /**
+     * State whether scroll position reaches the top
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:10]:
```suggestion
     * State whether the scroll position reaches the top.
```

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    public static boolean isHomeScrolledTop = false;
+
+    /**
+     * Detect event when back button is pressed
+     *
+     * @param activity is used when closing the app
+     */
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:18]:
```suggestion
    /**
     * Handle the event of clicking the back button.
     *
     * @param activity The activity, the app is launched with to finish.
     */
```

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    public static void exitOnBackPressed(WatchWhileActivity activity) {
+        // Check scroll position reaches the top in home feed
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:20]:
```suggestion
```

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        // Close the current activity once counter reached 1
+        if (backPressedCount <= 0) {
+            LogHelper.printDebug(() -> "BackPressed detected");
+
+            backPressedCount++;
+        } else {
+            LogHelper.printDebug(() -> "Activity is finished");
+
+            activity.finish();
+        }
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:32]: This makes no sense. Why is this check required? The activity should be finished when `isHomeScrolledTop` is true immediately, when `exitOnBackPressed` is called.

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    /**
+     * Detect event when ScrollView is created by RecyclerView
+     */
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:37]:
```suggestion
    /**
     * Handle the event, when it is being scrolled.
     */
```

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        isHomeScrolledTop = false;
+
+        backPressedCount = 0;
+
+        LogHelper.printDebug(() -> "ScrollView is created");
+    }
+
+    /**
+     * Detect event when the scroll position reaches the top by the back button
+     */
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:48]:
```suggestion
    /**
     * Handle the event of stopping to scroll.
     */
```

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    public static void onStopScrollView() {
+        isHomeScrolledTop = true;
+
+        LogHelper.printDebug(() -> "ScrollView is stopped");
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:52]:
```suggestion
        LogHelper.printDebug(() -> "Stopped scrolling");
```

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Detect event when ScrollView is created by RecyclerView
+     */
+    public static void onStartScrollView() {
+        isHomeScrolledTop = false;
+
+        backPressedCount = 0;
+
+        LogHelper.printDebug(() -> "ScrollView is created");
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:43]:
```suggestion
        LogHelper.printDebug(() -> "Started scrolling");
```

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
+package app.revanced.integrations.patches;
+
+import com.google.android.apps.youtube.app.watchwhile.WatchWhileActivity;
+import app.revanced.integrations.utils.LogHelper;
+
+public class BackToExitPatch {
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:6]: This class name should be the same as the counterpart patch class.

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * Handle the event of clicking the back button.
+     *
+     * @param activity The activity, the app is launched with to finish.
+     */
+    public static void exitOnBackPressed(WatchWhileActivity activity) {
+        if (!isHomeScrolledTop) return;
+
+        // Close the current activity
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:20]:
```suggestion
```

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        activity.finish();
+    }
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:22]:
```suggestion
        
        LogHelper.printDebug(() -> "Closed activity");
    }
```

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    public static void onStartScrollView() {
+        isHomeScrolledTop = false;
+
+        LogHelper.printDebug(() -> "Started scrolling");
+    }
+
+    /**
+     * Handle the event of stopping to scroll.
+     */
+    public static void onStopScrollView() {
```
Comment (oSumAtrIX) [inline app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java:37]: From what the code and comments convey, is that when scrolling stops, the next click on the back button will close the app. This should not be the case. Instead it should only close, when - as the variable name `isHomeScrolledTop` suggests - the feed is scrolled to top.

Comment (0xrxL) [reply app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java]: So what you suggest? To change the variables/methods names?

Comment (0xrxL) [reply app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java]: Anyway...the current code close the activity immediately after the list is entirely scrolled to the up (until the first video).

So should be good.

Comment (oSumAtrIX) [reply app/src/main/java/app/revanced/integrations/patches/BackToExitPatch.java]: > So what you suggest? To change the variables/methods names?

If the behavior works as intended, the comment and method name should be adjusted to reflect what they really do.

**Pull request merged**